src/app/spreadsheet.service.ts

```diff
@@ -1,34 +1,138 @@
 import { HttpClient } from '@angular/common/http';
 import { Injectable } from '@angular/core';
+import * as moment from 'moment';
 import { firstValueFrom, map, Observable, Subject } from 'rxjs';
+import { getBackendUrl, getGoogleClientID, getSpreadsheetID } from 'src/main';
 import { Note, Student } from 'src/types/student';
 
 const MS_IN_SECOND = 1000;
 
+export interface ISpreadsheetService {
+  /**
+   * Gets a list of all student names from the spreadsheet.
+   * 
+   * @returns An array of student names. 
+   * 
+   */
+  getStudentNames(): Promise<Observable<string[]>>;
+
+  /**
+   * Creates a new student on the spreadsheet.
+   * 
+   * @param student The student object to persist onto the spreadsheet. 
+   */
+  createStudent(student: Student): Promise<[Observable<Object>, Observable<Object>]>;
+
+  /**
+   * Gets all details about a student from the spreadsheet including their name, tracker, starting rank and notes.
+   * 
+   * @param studentName The name of the student in the spreadsheet.
+   */
+  getStudent(studentName: string): Promise<Observable<Student>>;
+
+  /**
+   * Updates data about a student in the spreadsheet.
+   * 
+   * @param student The updated student object to persist.
+   */
+  updateStudent(student: Student): Promise<[Observable<Object>, Observable<Object>]>;
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class SpreadsheetService {
+  constructor (
+    private spreadsheetReaderService: SpreadsheetReaderService,
+    private spreadsheetEditorService: SpreadsheetEditorService
+  ) { }
+
+  public get instance(): ISpreadsheetService {
+    if (this.spreadsheetEditorService.isLoggedIn()) return this.spreadsheetEditorService;
+    else return this.spreadsheetReaderService;
+  }
+
+  public async login(): Promise<void> {
+      await this.spreadsheetEditorService.login();
+  }
+
+  public isReadOnly() {
+    return this.spreadsheetEditorService.isLoggedIn();
+  }
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+export class SpreadsheetReaderService implements ISpreadsheetService {
+  constructor(
+    private httpClient: HttpClient
+  ) { }
+
+  private buildGetStudentNamesUrl(): string {
+    return `http://${getBackendUrl()}/list_students`;
+  }
+
+  private buildGetStudentUrl(studentName: string): string {
+    return `http://${getBackendUrl()}/student?name=${studentName}`;
+  }
```

Comment:
The `http` should not be hardcoded here so that `https` can be used for prod. (Also when I test locally I test with a separate API domain like prod will have, so it makes it easier for my local dev)